### PR TITLE
Csp 255 add import audit

### PR DESCRIPTION
**Pull request by MarkFCain**

### CODE REVIEW

File: src/SFA.DAS.Roatp.Data/Configuration/ImportAudit.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        public void Configure(EntityTypeBuilder<ImportAudit> builder)
+        {
+            builder.ToTable(nameof(ImportAudit));
+            builder.HasKey(p => p.Id);
+            builder.Property(p => p.ImportType).IsRequired();
+            builder.Property(p => p.RowsImported).IsRequired();
+            builder.Property(p => p.TimeStarted).IsRequired();
+            builder.Property(p => p.TimeFinished).IsRequired();
+            builder.Property(e => e.ImportType)
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Data/Configuration/ImportAudit.cs:20]: Import type is mentioned twice, you can merge all the rules together

File: src/SFA.DAS.Roatp.Data/Repositories/ImportAuditInsertRepository.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading.Tasks;
+using SFA.DAS.Roatp.Domain.Entities;
+using SFA.DAS.Roatp.Domain.Interfaces;
+
+namespace SFA.DAS.Roatp.Data.Repositories
+{
+    [ExcludeFromCodeCoverage]
+    internal class ImportAuditInsertRepository : IImportAuditInsertRepository
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Data/Repositories/ImportAuditInsertRepository.cs:10]: Rename to ImportAuditWriteRepository

File: src/SFA.DAS.Roatp.Data/Repositories/ImportAuditReadRepository.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    internal class ImportAuditReadRepository : IImportAuditReadRepository
+    {
+        private readonly RoatpDataContext _roatpDataContext;
+
+        public ImportAuditReadRepository(RoatpDataContext roatpDataContext)
+        {
+            _roatpDataContext = roatpDataContext;
+        }
+
+        public ImportAudit GetLastImportByType(string importType)
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Data/Repositories/ImportAuditReadRepository.cs:20]: What is the purpose of this method when we have another async method doing exactly the same?

File: src/SFA.DAS.Roatp.Database/Tables/ImportAudit.sql [hunk 4]
```diff
@@ -0,0 +1,9 @@
+﻿CREATE TABLE [dbo].[ImportAudit]
+(
+	[Id] INT PRIMARY KEY IDENTITY(1,1),
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Database/Tables/ImportAudit.sql:3]: Add explicit declaration of primary key with name

File: src/SFA.DAS.Roatp.Domain/Entities/ImportAudit.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+
+namespace SFA.DAS.Roatp.Domain.Entities
+{
+    public class ImportAudit
+    {
+        public ImportAudit(DateTime timeStarted, int rowsImported, ImportType importType)
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Domain/Entities/ImportAudit.cs:7]: Potentially bad design here, either do not assign TimeFinished or take it in the argument and default it.

File: src/SFA.DAS.Roatp.Jobs/Services/CourseDirectory/LoadCourseDirectoryDataService.cs [hunk 6]
```diff
@@ -91,7 +96,9 @@ public async Task<CourseDirectoryImportMetrics> LoadCourseDirectoryData(bool bet
                     }
                 }
             }
-            
+
+            await _importAuditInsertRepository.Insert(new ImportAudit(timeStarted, loadMetrics.NumberOfProvidersLoadedSuccessfully, ImportType.CourseDirectory));
+            _logger.LogInformation("Course Directory audit record written");
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Jobs/Services/CourseDirectory/LoadCourseDirectoryDataService.cs:101]: This log here does not make any sense

File: src/SFA.DAS.Roatp.Jobs/Services/ReloadStandardsCacheService.cs [hunk 7]
```diff
@@ -35,6 +39,9 @@ public async Task ReloadStandardsCache()
             await _reloadStandardsRepository.ReloadStandards(standardsToReload);
 
             _logger.LogInformation("Standards reload complete");
+            await _importAuditInsertRepository.Insert(new ImportAudit(timeStarted, standardsToReload.Count, ImportType.Standards));
+            _logger.LogInformation("Standards reload audit record written");
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Jobs/Services/ReloadStandardsCacheService.cs:43]: Again the log does not make sense, if you want to log, add the log to the repository so we dont have to repeat this every where.

File: src/SFA.DAS.Roatp.Data/Repositories/ImportAuditReadRepository.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            return  _roatpDataContext
+                    .ImportAudits
+                    .AsEnumerable()
+                    .OrderByDescending(c => c.TimeStarted)
+                    .Where(c => c.ImportType.ToString() == importType).ToList().FirstOrDefault();
+        }
+
+        public async Task<ImportAudit> GetLastImportByType(ImportType importType)
```
Comment (DevenKShah) [inline src/SFA.DAS.Roatp.Data/Repositories/ImportAuditReadRepository.cs:29]: I think we will ever only need the TimeStarted, so perhaps make this method more specific like `GetLastImportDateByImportType` and return `TimeStarted`. 

Also worth checking the SQL formed as you are ordering first and then filtering. I would rather ...

`ImportAudits.Where(ImportType).OrderByDesc(TimeStarted).Select(TimeStarted).FirstOrDefaultAsync()`

**Review by DevenKShah — COMMENTED** [review]
A few comments.

File: src/SFA.DAS.Roatp.Data/Repositories/ImportAuditReadRepository.cs (hunk 3 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Data/Repositories/ImportAuditReadRepository.cs]: I thought about it, and decided to take it out

File: src/SFA.DAS.Roatp.Data/Repositories/ImportAuditInsertRepository.cs (hunk 2 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Data/Repositories/ImportAuditInsertRepository.cs:10]: ok

File: src/SFA.DAS.Roatp.Data/Configuration/ImportAudit.cs (hunk 1 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Data/Configuration/ImportAudit.cs:20]: ok

File: src/SFA.DAS.Roatp.Database/Tables/ImportAudit.sql (hunk 4 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Database/Tables/ImportAudit.sql:3]: ok

File: src/SFA.DAS.Roatp.Jobs/Services/ReloadStandardsCacheService.cs (hunk 7 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Jobs/Services/ReloadStandardsCacheService.cs:43]: ok

File: src/SFA.DAS.Roatp.Jobs/Services/CourseDirectory/LoadCourseDirectoryDataService.cs (hunk 6 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Jobs/Services/CourseDirectory/LoadCourseDirectoryDataService.cs:101]: ok

File: src/SFA.DAS.Roatp.Data/Repositories/ImportAuditReadRepository.cs (hunk 8 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Data/Repositories/ImportAuditReadRepository.cs:20]: ok, will do

File: src/SFA.DAS.Roatp.Domain/Entities/ImportAudit.cs (hunk 5 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Domain/Entities/ImportAudit.cs:7]: I have taken out TimeFinished assignments, and made utcdate the default in the table

Comment (DevenKShah) [reply src/SFA.DAS.Roatp.Domain/Entities/ImportAudit.cs:7]: I will disagree to that as well. I think it better be assigned from the API so it is visible and also we can have tests around it.

**Review by DevenKShah — APPROVED** [review]
Looks good, I think the end time must be assigned specifically so it is not hidden in the table.

File: src/SFA.DAS.Roatp.Domain/Entities/ImportAudit.cs (hunk 5 above)
Comment (MarkFCain) [reply src/SFA.DAS.Roatp.Domain/Entities/ImportAudit.cs:7]: I'm note sure what you mean when you say 'assigned from the API'.  I've changed it so it's a default value if no parameter entered, but it overridable if you want to enter an explicit TimeFinished.

**Pull request merged**